compiler/src/iree/compiler/Dialect/Stream/Transforms/FuseDispatchBindings.cpp

```diff
@@ -322,6 +322,25 @@ fuseDispatchBindings(IREE::Stream::ExecutableOp executableOp,
   auto anyDispatchOp = dispatchOps.front();
   unsigned bindingCount = anyDispatchOp.getResources().size();
 
+  // Bail if we can't find an internal function reference to update.
+  auto funcOp = exportOp.lookupFunctionRef();
+  if (!funcOp)
+    return;
+
+  int64_t numBindingsOnFunc = 0;
+  for (auto arg : funcOp.front().getArguments()) {
+    if (llvm::isa<IREE::Stream::BindingType>(arg.getType())) {
+      numBindingsOnFunc++;
+    }
+  }
+  // Bail if the exported function does not have the same number of bindings
+  // as the dispatch. This could be because the executable already lowered the
+  // bindings, in which case we can no longer reliably fuse them. We could try
+  // to be clever and fuse a subset of bindings, but this typically only happens
+  // when the user is trying to do something custom.
+  if (numBindingsOnFunc != bindingCount)
```

Comment:
this seems a bit too complex - are there any cases where we'd have non-zero partial bindings?
(would rather keep this simple: if there are bindings we optimize them, if there aren't we don't)

Reply:
The only case I can think of would be if a user decided to write it like that, so maybe not worth ever supporting.

Reply:
it'd also be good to understand what you're doing here - we really shouldn't be use stream ops without bindings and instead generate hal ops directly - the only reason to use stream ops is to get the stuff stream dialect can do

Reply:
I was doing this: https://gist.github.com/qedawkins/1f344eb7551d063e360ce9d3a85ab0a7

Using a `flow.executable`. My basic thought process was if I wanted to use mostly the same input IR for CPU and SPIRV, I should go through Flow ops instead of having to create variants with specific targets (even though in the example I put SPIR-V codegen configs).

If this IR shouldn't be allowed, then we could strengthen the verifier of `flow.executable`, else it will turn into stream IR and then crash here without this change.

Reply:
yeah, that's a misuse - flow/stream are expecting flow/stream stuff, not hal - you don't need to specify hal bindings and such - just use the appropriate flow ops.

we can't verify that users don't put in random things - we allow custom ops and such - from the flow/stream perspective the hal is just some random other dialect

Reply:
ah `flow.dispatch.workgroup.id`, let me see how that works.

Reply:
yeah give that a go and if there's stuff missing we can add additional plumbing/routes through - I'm pretty sure we don't want to support this kind of usage (using hal ops with flow executables) so I'm not sure we want to land this extra logic (there shouldn't be cases where there are flow/stream executables with bodies that we can't touch - lots of passes aren't expecting that)

Reply:
Makes sense, I guess it was kind of hard to tell that this was unexpected usage, at least when I was comparing against external objects on `flow.executables` (the precursor to `hal.dispatch.extern`). It seemed like we were already doing this for those cases, and doing the same for something with an explicit body that we expect some kind of codegen+target to handle didn't seem very different. Also this was the only pass to crash :P

(side note: really wish Github didn't require a reload to get messages in conversations like this. Makes it hard to track responses)

Reply:
nice, that worked. I can either close this or change it to signal a pass failure with some kind of error message when the function signature doesn't include enough bindings (in case a user tried to do something like this in the future).

Reply:
verifying that binding count matches would be awesome - AsyncDispatchOp/CmdDispatchOp have a TODO for "verify that the target function has matching operands" that'd be good to resolve - it currently just checks the workload -> workgroup count region

Reply:
I just did it for CmdDispatchOp for now, but it might be easier to just do both at the same time hmmm